### PR TITLE
Introduce initial API scaffolding and modules.

### PR DESCRIPTION
**Pull request by xavierh**
Back port from 1.11

review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java [hunk 1]
```diff
@@ -95,10 +98,10 @@ public void toBytes(@NotNull final ByteBuf buf)
      */
     public static class Permission extends AbstractMessage<Permission, IMessage>
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java:99]: MAJOR Reduce this class from 63 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 2]
```diff
@@ -437,6 +443,35 @@ public WindowTownHall(final BuildingTownHall.View townHall)
...
+    }
+
+    /**
+     * Clears and resets all citizens.
+     */
+    private void updateWorkOrders()
+    {
+        workOrders.clear();
+        workOrders.addAll(townHall.getColony().getWorkOrders());
+        workOrders.sort((first, second) -> second.getPriority() > first.getPriority() ? 1 : (second.getPriority() < first.getPriority() ? -1 : 0));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:472]: MAJOR Extract this nested ternary operation into an independent statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3358)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 3]
```diff
@@ -800,7 +806,7 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
                 }
 
                 rowPane.findPaneOfTypeByID("name", Label.class).setLabelText(name);
-                final boolean isTriggered = townHall.getColony().getPermissions().hasPermission(Permissions.Rank.valueOf(actionsList.getParent().getID().toUpperCase()), action);
+                final boolean isTriggered = townHall.getColony().getPermissions().hasPermission(Rank.valueOf(actionsList.getParent().getID().toUpperCase()), action);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:809]: MINOR Define the locale to be used in this String operation. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1449)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityColonyBuilding.java [hunk 4]
```diff
@@ -290,7 +290,7 @@ public boolean isUseableByPlayer(@NotNull final EntityPlayer player)
     public boolean hasAccessPermission(final EntityPlayer player)
     {
         //TODO This is called every tick the GUI is open. Is that bad?
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityColonyBuilding.java:292]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 106 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 51 major
* ![MINOR][MINOR] 40 minor
* ![INFO][INFO] 14 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [InventoryCitizen.java#L347](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java#L347): Reduce the number of conditional operators (7) used in the expression (maximum allowed 6). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L658](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L658): Extract this nested ternary operation into an independent statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3358)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L788](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L788): Reduce this anonymous class number of lines from 24 to at most 20, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L885](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L885): Reduce this anonymous class number of lines from 40 to at most 20, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.238 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L44](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java#L44): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 43 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L44](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java#L44): class "Colony" has 83 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L642](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java#L642): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1164](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1164): Assign this magic number 20 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1165](https://github.com/Minecolonies/minecolonies/blob/5b1d78c20be166d20a71fe067fea084666eed9d9/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1165): Assign this magic number 60 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1068/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Review by marchermans — APPROVED** [review]

**Pull request merged**